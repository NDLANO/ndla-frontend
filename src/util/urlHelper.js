/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { matchPath } from 'react-router-dom';
import parseUrl from 'parse-url';
import { isValidLocale } from '../i18n';
import {
  RESOURCE_PAGE_PATH,
  PLAIN_ARTICLE_PAGE_PATH,
  TOPIC_PATH,
} from '../constants';

function matchUrl(pathname, type, lang = false) {
  if (type === 'article') {
    return matchPath(
      pathname,
      lang ? `/:lang${PLAIN_ARTICLE_PAGE_PATH}` : PLAIN_ARTICLE_PAGE_PATH,
    );
  }
  let match = matchPath(
    pathname,
    lang ? `/:lang${RESOURCE_PAGE_PATH}` : RESOURCE_PAGE_PATH,
  );
  console.log('nfhfhr', match);
  if (match) {
    return match;
  }

  return matchPath(pathname, lang ? `/:lang${TOPIC_PATH}` : TOPIC_PATH);
}

export function parseAndMatchUrl(url) {
  const { pathname } = parseUrl(url);
  const paths = pathname.split('/');
  paths[1] = paths[1] === 'unknown' ? 'nb' : paths[1];
  const path = paths.join('/');

  if (isValidLocale(paths[1])) {
    return matchUrl(path, paths[2], true);
  }
  return matchUrl(path, paths[1], false);
}
