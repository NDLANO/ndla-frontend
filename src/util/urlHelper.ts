/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { matchPath, Params, PathMatch, Location } from "react-router-dom";
import { GQLTaxBase } from "../graphqlTypes";
import { isValidLocale, supportedLanguages } from "../i18n";
import { oembedRoutes } from "../routes";

type OembedParams =
  | "subjectId"
  | "topicId"
  | "resourceId"
  | "stepId"
  | "contextId"
  | "articleId"
  | "lang"
  | "topicOrResourceId"
  | "audioId"
  | "videoId"
  | "imageId"
  | "conceptId"
  | "h5pId";

type OembedReturnParams =
  | "subjectId"
  | "topicId"
  | "resourceId"
  | "stepId"
  | "contextId"
  | "articleId"
  | "lang"
  | "audioId"
  | "videoId"
  | "imageId"
  | "conceptId"
  | "h5pId";

const matchUrl = (pathname: string, lang: boolean = false): PathMatch<OembedReturnParams> | null => {
  const possiblePaths = lang ? oembedRoutes.map((r) => `/:lang/${r}`) : oembedRoutes;

  let match: PathMatch<OembedParams> | undefined;
  for (let i = 0; i < possiblePaths.length; i++) {
    const attempt = possiblePaths[i] ?? "";
    const m = matchPath<OembedParams, string>(attempt, pathname);
    if (m) {
      match = m;
      break;
    }
  }
  if (!match) {
    return null;
  }

  if (match.params.topicOrResourceId) {
    const missingParam: {
      resourceId?: string;
      topicId?: string;
    } = match.params.topicOrResourceId.startsWith(":resource")
      ? { resourceId: match.params.topicOrResourceId.replace(":resource", "") }
      : { topicId: match.params.topicOrResourceId.replace(":topic", "") };
    const params: Params<OembedParams> = match.params;
    return {
      ...match,
      params: {
        ...params,
        ...missingParam,
      },
    };
  } else return match;
};

export function parseOembedUrl(url: string, ignoreLocale: boolean = false) {
  const { pathname } = new URL(url);
  const paths = pathname.split("/");
  if (paths[1]) {
    paths[1] = paths[1] === "unknown" ? "nb" : paths[1];
  }
  if (ignoreLocale && isValidLocale(paths[1])) {
    paths.splice(1, 1);
  }
  const path = paths.join("/");

  if (isValidLocale(paths[1])) {
    return matchUrl(path, true);
  }
  return matchUrl(path, false);
}

export const toHref = (location: Location) => {
  return `${location.pathname}${location.search}`;
};

export const constructNewPath = (pathname: string, newLocale?: string) => {
  const regex = new RegExp(`\\/(${supportedLanguages.join("|")})($|\\/)`, "");
  const path = pathname.replace(regex, "");
  const fullPath = path.startsWith("/") ? path : `/${path}`;
  const localePrefix = newLocale ? `/${newLocale}` : "";
  return `${localePrefix}${fullPath}`;
};

export const isCurrentPage = (pathname: string, taxBase: Pick<GQLTaxBase, "path" | "url">) => {
  let path = pathname.replace(/\/$/, ""); // Remove trailing slash if present
  const match = matchUrl(path);
  if (match?.params.stepId) {
    path = path.replace(/\/\d+$/, ""); // Remove last numeric segment if stepId
  }
  return path === taxBase.path || decodeURIComponent(path) === taxBase.url;
};
