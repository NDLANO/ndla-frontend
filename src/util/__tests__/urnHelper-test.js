/**
 * Copyright (c) 2019-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { parseAndMatchUrl } from '../urlHelper';

const validArticleUrl1 =
  'https://ndla-frontend.test.api.ndla.no/subjects/subject:3/topic:1:55163/topic:1:168398/resource:1:1682';
const validArticleUrl2 =
  'http://localhost:3000/subjects/subject:3/topic:1:168542/topic:1:173292/resource:1:168554';
const validArticleUrl3 =
  'http://localhost:3000/subjects/subject:3/topic:1:168542/resource:1:168554';
const validArticleUrl4 =
  'https://ndla-frontend.test.api.ndla.no/unknown/subjects/subject:3/topic:1:55163/topic:1:168398/resource:1:1682';

const validArticleUrlWithLang =
  'https://ndla-frontend.test.api.ndla.no/nn/subjects/subject:3/topic:1:55163/topic:1:168398/resource:1:1682';
const validTopicArticleUrl =
  'https://ndla-frontend.test.api.ndla.no/subjects/subject:3/topic:1:55163';
const unvalidArticleUrl =
  'https://ndla-frontend.test.api.ndla.no/subjects/subject:3';

const validateSimpleArticlePath1 =
  'https://ndla-frontend.test.api.ndla.no/article/4809';
const validateSimpleArticlePath2 =
  'https://ndla-frontend.test.api.ndla.no/nn/article/4809';

const unvalidateSimpleArticlePath =
  'https://ndla-frontend.test.api.ndla.no/articles/4809';

test('parseAndMatchUrl', () => {
  expect(parseAndMatchUrl(validArticleUrl1)).toMatchSnapshot();
  expect(parseAndMatchUrl(validTopicArticleUrl)).toMatchSnapshot();
  expect(parseAndMatchUrl(validArticleUrl2)).toMatchSnapshot();
  expect(parseAndMatchUrl(validArticleUrl3)).toMatchSnapshot();
  expect(parseAndMatchUrl(validArticleUrl4)).toMatchSnapshot();
  expect(parseAndMatchUrl(validArticleUrlWithLang)).toMatchSnapshot();
  expect(parseAndMatchUrl(unvalidArticleUrl)).toBe(null);
  expect(parseAndMatchUrl(validateSimpleArticlePath1)).toMatchSnapshot();
  expect(parseAndMatchUrl(validateSimpleArticlePath2)).toMatchSnapshot();
  expect(parseAndMatchUrl(unvalidateSimpleArticlePath)).toBe(null);
});
