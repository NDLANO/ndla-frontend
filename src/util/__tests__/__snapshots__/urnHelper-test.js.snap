// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseAndMatchUrl 1`] = `
Object {
  "isExact": true,
  "params": Object {
    "resourceId": "1:1682",
    "stepId": undefined,
    "subjectId": "subject:3",
    "topicId": "topic:1:168398",
    "topicPath": "topic:1:55163",
  },
  "path": "/subjects/:subjectId/:topicPath*/:topicId/resource\\\\::resourceId/:stepId?",
  "url": "/subjects/subject:3/topic:1:55163/topic:1:168398/resource:1:1682",
}
`;

exports[`parseAndMatchUrl 2`] = `
Object {
  "isExact": true,
  "params": Object {
    "subjectId": "subject:3",
    "topicId": "topic:1:55163",
    "topicPath": undefined,
  },
  "path": "/subjects/:subjectId/:topicPath(.*)?/:topicId",
  "url": "/subjects/subject:3/topic:1:55163",
}
`;

exports[`parseAndMatchUrl 3`] = `
Object {
  "isExact": true,
  "params": Object {
    "resourceId": "1:168554",
    "stepId": undefined,
    "subjectId": "subject:3",
    "topicId": "topic:1:173292",
    "topicPath": "topic:1:168542",
  },
  "path": "/subjects/:subjectId/:topicPath*/:topicId/resource\\\\::resourceId/:stepId?",
  "url": "/subjects/subject:3/topic:1:168542/topic:1:173292/resource:1:168554",
}
`;

exports[`parseAndMatchUrl 4`] = `
Object {
  "isExact": true,
  "params": Object {
    "resourceId": "1:168554",
    "stepId": undefined,
    "subjectId": "subject:3",
    "topicId": "topic:1:168542",
    "topicPath": undefined,
  },
  "path": "/subjects/:subjectId/:topicPath*/:topicId/resource\\\\::resourceId/:stepId?",
  "url": "/subjects/subject:3/topic:1:168542/resource:1:168554",
}
`;

exports[`parseAndMatchUrl 5`] = `
Object {
  "isExact": true,
  "params": Object {
    "lang": "nb",
    "resourceId": "1:1682",
    "stepId": undefined,
    "subjectId": "subject:3",
    "topicId": "topic:1:168398",
    "topicPath": "topic:1:55163",
  },
  "path": "/:lang/subjects/:subjectId/:topicPath*/:topicId/resource\\\\::resourceId/:stepId?",
  "url": "/nb/subjects/subject:3/topic:1:55163/topic:1:168398/resource:1:1682",
}
`;

exports[`parseAndMatchUrl 6`] = `
Object {
  "isExact": true,
  "params": Object {
    "lang": "nn",
    "resourceId": "1:1682",
    "stepId": undefined,
    "subjectId": "subject:3",
    "topicId": "topic:1:168398",
    "topicPath": "topic:1:55163",
  },
  "path": "/:lang/subjects/:subjectId/:topicPath*/:topicId/resource\\\\::resourceId/:stepId?",
  "url": "/nn/subjects/subject:3/topic:1:55163/topic:1:168398/resource:1:1682",
}
`;

exports[`parseAndMatchUrl 7`] = `
Object {
  "isExact": true,
  "params": Object {
    "articleId": "4809",
  },
  "path": "/article/:articleId",
  "url": "/article/4809",
}
`;

exports[`parseAndMatchUrl 8`] = `
Object {
  "isExact": true,
  "params": Object {
    "articleId": "4809",
    "lang": "nn",
  },
  "path": "/:lang/article/:articleId",
  "url": "/nn/article/4809",
}
`;
